README.txt
= timecode

* http://guerilla-di.rubyforge.org/timecode

== DESCRIPTION:

Value class for SMPTE timecode information

== SYNOPSIS:

  tc = Timecode.parse("00:00:10:12", fps = 25)
  tc.total #=> 262
  
  plus_ten = tc + Timecode.parse("10h", fps = 25)
  plus_ten #=> "10:00:10:12"
  
== PROBLEMS:

Currently there is no support for drop-frame timecode

== INSTALL:

* sudo gem install timecode

== LICENSE:

(The MIT License)

Copyright (c) 2008 Julik Tarkhanov

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.2.0 / 2009-02-26

* Properly handle integers starting with 0 in Tiecode.parse (zef)

=== 0.1.9 / 2009-02-23

* Do not be biased when rounding out frames on from_seconds

=== 0.1.8 / 2009-01-31

* Bail out with empty or whitespace strings passed to Timecode.parse
* Add Timecode#to_seconds that returns a float value (useful for Quicktime)
* Move the validation routine into the class itself

=== 0.1.7 / 2009-01-31

* Simplified parsing code, more safeguards and needless exceptions removed

=== 0.1.6 / 2009-01-28

* Fixed RDoc linkage
* Custom Timecode descendants should hold on to their class

=== 0.1.5 / 2009-01-14

* Tests remade using test/spec

=== 0.1.4 / 2000-01-01

* Expanded test coverage
* Some formatting/doc improvements

=== 0.1.3 / 2008-12-25

* Implement the format FFmpeg uses (fractional seconds instead of frames)

=== 0.1.2 / 2008-12-25

* Fix to_uint
* Always use float frame rates and rely on a small delta for comparison

=== 0.1.1 / 2008-12-15

* Allow passing framerate to from_uint

=== 0.1.0 / 2008-12-15

* 1 major enhancement

  * Birthday!



test_timecode.rb
require 'test/unit'
require 'rubygems'
require 'timecode'

# for Fixnum#hours
require 'active_support'

class TimecodeTest < Test::Unit::TestCase
  
  def test_basics
    five_seconds_of_pal = 5.seconds * 25
    tc = Timecode.new(five_seconds_of_pal, 25)
    assert_equal 0, tc.hours
    assert_equal 0, tc.minutes
    assert_equal 5, tc.seconds
    assert_equal 0, tc.frames
    assert_equal five_seconds_of_pal, tc.total
    assert_equal "00:00:05:00", tc.to_s
    
    one_and_a_half_hour_of_hollywood = 90.minutes * 24

    film_tc = Timecode.new(one_and_a_half_hour_of_hollywood, 24)
    assert_equal 1, film_tc.hours
    assert_equal 30, film_tc.minutes
    assert_equal 0, film_tc.seconds
    assert_equal 0, film_tc.frames
    assert_equal one_and_a_half_hour_of_hollywood, film_tc.total
    assert_equal "01:30:00:00", film_tc.to_s

    assert_equal "01:30:00:04", (film_tc + 4).to_s
    assert_equal "01:30:01:04", (film_tc + 28).to_s

    assert_raise(Timecode::WrongFramerate) do
      tc + film_tc
    end
    
    two_seconds_and_five_frames_of_pal = ((2.seconds * 25) + 5)
    pal_tc = Timecode.new(two_seconds_and_five_frames_of_pal, 25)
    assert_nothing_raised do
      added_tc = pal_tc + tc
      assert_equal "00:00:07:05", added_tc.to_s
    end

  end
    
  def test_parse
    simple_tc = "00:10:34:10"
    
    assert_nothing_raised do
      @tc = Timecode.parse(simple_tc)
      assert_equal simple_tc, @tc.to_s
    end
  
    bad_tc = "00:76:89:30"
    unknown_gobbledygook = "this is insane"
    
    assert_raise(Timecode::CannotParse) do
      tc = Timecode.parse(unknown_gobbledygook, 25)
    end
    
    assert_raise(Timecode::RangeError) do
      Timecode.parse(bad_tc, 25)
    end
  end
  
  def test_succ
    assert_equal Timecode.new(23), Timecode.new(22).succ
  end
  
  def test_zero
    assert Timecode.new(0).zero?
    assert !Timecode.new(1).zero?
    assert !Timecode.new(1000).zero?
  end
  
  def test_parse_from_numbers
    assert_equal Timecode.new(10), Timecode.parse("10")
    assert_equal Timecode.new(60), Timecode.parse("210")
    assert_equal "10:10:10:10", Timecode.parse("10101010").to_s
  end
  
  def test_parse_with_f
    assert_equal Timecode.new(60), Timecode.parse("60f")
  end
  
  def test_parse_s
    assert_equal Timecode.new(50), Timecode.parse("2s")
    assert_equal Timecode.new(60), Timecode.parse("2s", 30)
  end

  def test_parse_m
    assert_equal Timecode.new(25 * 60 * 3), Timecode.parse("3m")
  end

  def test_parse_h
    assert_equal Timecode.new(25 * 60 * 60 * 3), Timecode.parse("3h")
  end
  
  def test_parse_block
    assert_equal '01:00:00:04', Timecode.parse("1h 4f").to_s
    assert_equal '01:00:00:04', Timecode.parse("4f 1h").to_s
    assert_equal '01:00:01:04', Timecode.parse("29f 1h").to_s
  end
  
  def test_float_framerate
    tc = Timecode.new(25, 12.5)
    assert_equal "00:00:02:00", tc.to_s
  end
  
  def test_timecode_with_nil_gives_zero
    assert_equal Timecode.new(0), Timecode.new(nil)
  end
  
  def test_parse_fractional_tc
    fraction = "00:00:07.1"
    tc = Timecode.parse_with_fractional_seconds(fraction, 10)
    assert_equal "00:00:07:01", tc.to_s

    fraction = "00:00:07.5"
    tc = Timecode.parse_with_fractional_seconds(fraction, 10)
    assert_equal "00:00:07:05", tc.to_s
    
    fraction = "00:00:07.04"
    tc = Timecode.parse_with_fractional_seconds(fraction, 12.5)
    assert_equal "00:00:07:00", tc.to_s
    
    fraction = "00:00:07.16"
    tc = Timecode.parse_with_fractional_seconds(fraction, 12.5)
    assert_equal "00:00:07:02", tc.to_s
  end
  
  def test_parse_with_calculation
    tc = Timecode.parse_with_calculation("00:00:00:15 +2f")
    assert_equal Timecode.new(17), tc
  end
  
  def test_from_seconds
    fraction = 7.1
    tc = Timecode.from_seconds(fraction, 10)
    assert_equal "00:00:07:01", tc.to_s

    fraction = 7.5
    tc = Timecode.from_seconds(fraction, 10)
    assert_equal "00:00:07:05", tc.to_s

    fraction = 7.16
    tc = Timecode.from_seconds(fraction, 12.5)
    assert_equal "00:00:07:02", tc.to_s
  end
end


timecode_test.rb
require 'test/unit'
require 'rubygems'
require 'active_support' # for Fixnum#hours

class TimecodeTest < Test::Unit::TestCase
  
  def test_basics
    five_seconds_of_pal = 5.seconds * 25
    tc = Timecode.new(five_seconds_of_pal, 25)
    assert_equal 0, tc.hours
    assert_equal 0, tc.minutes
    assert_equal 5, tc.seconds
    assert_equal 0, tc.frames
    assert_equal five_seconds_of_pal, tc.total
    assert_equal "00:00:05:00", tc.to_s
    
    one_and_a_half_hour_of_hollywood = 90.minutes * 24

    film_tc = Timecode.new(one_and_a_half_hour_of_hollywood, 24)
    assert_equal 1, film_tc.hours
    assert_equal 30, film_tc.minutes
    assert_equal 0, film_tc.seconds
    assert_equal 0, film_tc.frames
    assert_equal one_and_a_half_hour_of_hollywood, film_tc.total
    assert_equal "01:30:00:00", film_tc.to_s
    
    assert_raise(Timecode::WrongFramerate) do
      tc + film_tc
    end
    
    two_seconds_and_five_frames_of_pal = ((2.seconds * 25) + 5)
    pal_tc = Timecode.new(two_seconds_and_five_frames_of_pal, 25)
    assert_nothing_raised do
      added_tc = pal_tc + tc
      assert_equal "00:00:07:05", added_tc.to_s
    end

  end
    
  def test_parse
    simple_tc = "00:10:34:10"
    
    assert_nothing_raised { @tc = Timecode.parse(simple_tc) }
  
    bad_tc = "00:76:89:30"
    
    assert_raise(Timecode::RangeError) do
      Timecode.parse(bad_tc, 25)
    end
  end
  
  def test_float_framerate
    tc = Timecode.new(25, 12.5)
    assert_equal "00:00:02:00", tc.to_s
  end
  
  def test_parse_fractional_tc
    fraction = "00:00:07.1"
    tc = Timecode.parse_with_fractional_seconds(fraction, 10)
    assert_equal "00:00:07:01", tc.to_s

    fraction = "00:00:07.5"
    tc = Timecode.parse_with_fractional_seconds(fraction, 10)
    assert_equal "00:00:07:05", tc.to_s
    
    fraction = "00:00:07.04"
    tc = Timecode.parse_with_fractional_seconds(fraction, 12.5)
    assert_equal "00:00:07:00", tc.to_s
    
    fraction = "00:00:07.16"
    tc = Timecode.parse_with_fractional_seconds(fraction, 12.5)
    assert_equal "00:00:07:02", tc.to_s
  end

  def test_from_seconds
    fraction = 7.1
    tc = Timecode.from_seconds(fraction, 10)
    assert_equal "00:00:07:01", tc.to_s

    fraction = 7.5
    tc = Timecode.from_seconds(fraction, 10)
    assert_equal "00:00:07:05", tc.to_s

    fraction = 7.16
    tc = Timecode.from_seconds(fraction, 12.5)
    assert_equal "00:00:07:02", tc.to_s
  end
end


lib/timecode.rb
# Timecode is a convenience object for calculating SMPTE timecode natively. 
# The promise is that you only have to store two values to know the timecode - the amount
# of frames and the framerate. An additional perk might be to save the dropframeness,
# but we avoid that at this point.
#
# You can calculate in timecode objects ass well as with conventional integers and floats.
# Timecode is immutable and can be used as a value object. Timecode objects are sortable.
#
# Here's how to use it with ActiveRecord (your column names will be source_tc_frames_total and tape_fps)
#
#   composed_of :source_tc, :class_name => 'Timecode',
#     :mapping => [%w(source_tc_frames total), %w(tape_fps fps)]

class Timecode
  VERSION = '0.2.0'

  include Comparable
  
  DEFAULT_FPS = 25.0
  
  #:stopdoc:
  NTSC_FPS = (30.0 * 1000 / 1001).freeze
  ALLOWED_FPS_DELTA = (0.001).freeze
  
  COMPLETE_TC_RE = /^(\d{2}):(\d{2}):(\d{2}):(\d{2})$/
  DF_TC_RE = /^(\d{1,2}):(\d{1,2}):(\d{1,2});(\d{2})$/
  FRACTIONAL_TC_RE = /^(\d{2}):(\d{2}):(\d{2}).(\d{1,8})$/
  
  WITH_FRACTIONS_OF_SECOND = "%02d:%02d:%02d.%02d"
  WITH_FRAMES = "%02d:%02d:%02d:%02d"
  #:startdoc:
  
  # All Timecode lib errors inherit from this
  class Error < RuntimeError; end
  
  # Gets raised if timecode is out of range (like 100 hours long)
  class RangeError < Error; end

  # Gets raised when a timecode cannot be parsed
  class CannotParse < Error; end

  # Gets raised when you try to compute two timecodes with different framerates together
  class WrongFramerate < ArgumentError; end
  
  # Initialize a new Timecode object with a certain amount of frames and a framerate
  # will be interpreted as the total number of frames
  def initialize(total = 0, fps = DEFAULT_FPS)
    raise WrongFramerate, "FPS cannot be zero" if fps.zero?
    
    # If total is a string, use parse
    raise RangeError, "Timecode cannot be negative" if total.to_i < 0
    # Always cast framerate to float, and num of rames to integer
    @total, @fps = total.to_i, fps.to_f
    @value = validate!
    freeze
  end
  
  def inspect # :nodoc:
    "#<Timecode:%s (%dF@%.2f)>" % [to_s, total, fps]
  end
  
  class << self
    
    # Use initialize for integers and parsing for strings
    def new(from, fps = DEFAULT_FPS)
      from.is_a?(String) ? parse(from, fps) : super(from, fps)
    end
    
    # Parse timecode and return zero if none matched
    def soft_parse(input, with_fps = DEFAULT_FPS)
      parse(input) rescue new(0, with_fps)
    end
      
    # Parse timecode entered by the user. Will raise if the string cannot be parsed.
    # The following formats are supported:
    # * 10h 20m 10s 1f (or any combination thereof) - will be disassembled to hours, frames, seconds and so on automatically
    # * 123 - will be parsed as 00:00:01:23
    # * 00:00:00:00 - will be parsed as zero TC
    def parse(input, with_fps = DEFAULT_FPS)
      # Drop frame goodbye
      if (input =~ DF_TC_RE)
        raise Error, "We do not support drop-frame TC"
      # 00:00:00:00
      elsif (input =~ COMPLETE_TC_RE)
        atoms_and_fps = input.scan(COMPLETE_TC_RE).to_a.flatten.map{|e| e.to_i} + [with_fps]
        return at(*atoms_and_fps)
      # 00:00:00.0
      elsif input =~ FRACTIONAL_TC_RE
        parse_with_fractional_seconds(input, with_fps)
      # 10h 20m 10s 1f 00:00:00:01 - space separated is a sum of parts
      elsif input =~ /\s/
        parts = input.gsub(/\s/, ' ').split.reject{|e| e.strip.empty? }
        raise CannotParse, "No atoms" if parts.empty?
        parts.map{|part|  parse(part, with_fps) }.inject{|sum, p| sum + p.total }
      # 10s
      elsif input =~ /^(\d+)s$/
        return new(input.to_i * with_fps, with_fps)
      # 10h
      elsif input =~ /^(\d+)h$/i
        return new(input.to_i * 60 * 60 * with_fps, with_fps)
      # 20m
      elsif input =~ /^(\d+)m$/i
        return new(input.to_i * 60 * with_fps, with_fps)
      # 60f - 60 frames, or 2 seconds and 10 frames
      elsif input =~ /^(\d+)f$/i
        return new(input.to_i, with_fps)
      # Only a bunch of digits, treat 12345 as 00:01:23:45
      elsif (input =~ /^(\d+)$/)
        atoms_len = 2 * 4
        # left-pad input AND truncate if needed
        padded = input[0..atoms_len].rjust(8, "0")
        atoms = padded.scan(/(\d{2})/).flatten.map{|e| e.to_i } + [with_fps]
        return at(*atoms)
      else
        raise CannotParse, "Cannot parse #{input} into timecode, unknown format"
      end
    end
    
    # Initialize a Timecode object at this specfic timecode
    def at(hrs, mins, secs, frames, with_fps = DEFAULT_FPS)
      validate_atoms!(hrs, mins, secs, frames, with_fps)
      total = (hrs*(60*60*with_fps) +  mins*(60*with_fps) + secs*with_fps + frames).round
      new(total, with_fps)
    end
    
    # Validate the passed atoms for the concrete framerate
    def validate_atoms!(hrs, mins, secs, frames, with_fps)
      case true
        when hrs > 99
          raise RangeError, "There can be no more than 99 hours, got #{hrs}"
        when mins > 59
          raise RangeError, "There can be no more than 59 minutes, got #{mins}"
        when secs > 59
          raise RangeError, "There can be no more than 59 seconds, got #{secs}"
        when frames > (with_fps -1)
          raise RangeError, "There can be no more than #{with_fps -1} frames @#{with_fps}, got #{frames}"
      end
    end
    
    # Parse a timecode with fractional seconds instead of frames. This is how ffmpeg reports
    # a timecode
    def parse_with_fractional_seconds(tc_with_fractions_of_second, fps = DEFAULT_FPS)
      fraction_expr = /\.(\d+)$/
      fraction_part = ('.' + tc_with_fractions_of_second.scan(fraction_expr)[0][0]).to_f

      seconds_per_frame = 1.0 / fps.to_f
      frame_idx = (fraction_part / seconds_per_frame).floor

      tc_with_frameno = tc_with_fractions_of_second.gsub(fraction_expr, ":%02d" % frame_idx)

      parse(tc_with_frameno, fps)
    end
  
    # create a timecode from the number of seconds. This is how current time is supplied by
    # QuickTime and other systems which have non-frame-based timescales
    def from_seconds(seconds_float, the_fps = DEFAULT_FPS)
      total_frames = (seconds_float.to_f * the_fps.to_f).to_i
      new(total_frames, the_fps)
    end
  
    # Some systems (like SGIs) and DPX format store timecode as unsigned integer, bit-packed. This method
    # unpacks such an integer into a timecode.
    def from_uint(uint, fps = DEFAULT_FPS)
      tc_elements = (0..7).to_a.reverse.map do | multiplier | 
        ((uint >> (multiplier * 4)) & 0x0F)
      end.join.scan(/(\d{2})/).flatten.map{|e| e.to_i}

      tc_elements << fps
      at(*tc_elements)
    end
  end
  
  # is the timecode at 00:00:00:00
  def zero?
    @total.zero?
  end
  
  # get total frame count
  def total
    to_f
  end
  
  # get FPS
  def fps
    @fps
  end
    
  # get the number of frames
  def frames
    value_parts[3]
  end
  
  # get the number of seconds
  def seconds
    value_parts[2]
  end
  
  # get the number of minutes
  def minutes
    value_parts[1]
  end
  
  # get the number of hours
  def hours
    value_parts[0]
  end
  
  # get frame interval in fractions of a second
  def frame_interval
    1.0/@fps
  end
  
  # get the timecode as bit-packed unsigned 32 bit int (suitable for DPX and SGI)
  def to_uint
    elements = (("%02d" * 4) % [hours,minutes,seconds,frames]).split(//).map{|e| e.to_i }
    uint = 0
    elements.reverse.each_with_index do | p, i |
      uint |= p << 4 * i 
    end
    uint
  end
  
  # get the timecode as a floating-point number of seconds (used in Quicktime)
  def to_seconds
    (@total / @fps)
  end
  
  # Convert to different framerate based on the total frames. Therefore,
  # 1 second of PAL video will convert to 25 frames of NTSC (this 
  # is suitable for PAL to film TC conversions and back).
  def convert(new_fps)
    self.class.new(@total, new_fps)
  end
  
  # get formatted SMPTE timecode
  def to_s
    WITH_FRAMES % value_parts
  end
  
  # get total frames as float
  def to_f
    @total
  end

  # get total frames as integer
  def to_i
    @total
  end
  
  # add number of frames (or another timecode) to this one
  def +(arg)
    if (arg.is_a?(Timecode) && framerate_in_delta(arg.fps, @fps))
      self.class.new(@total+arg.total, @fps)
    elsif (arg.is_a?(Timecode))
      raise WrongFramerate, "You are calculating timecodes with different framerates"
    else
      self.class.new(@total + arg, @fps)
    end
  end
  
  # Subtract a number of frames
  def -(arg)
    if (arg.is_a?(Timecode) &&  framerate_in_delta(arg.fps, @fps))
      self.class.new(@total-arg.total, @fps)
    elsif (arg.is_a?(Timecode))
      raise WrongFramerate, "You are calculating timecodes with different framerates"
    else
      self.class.new(@total-arg, @fps)
    end
  end
  
  # Multiply the timecode by a number
  def *(arg)
    raise RangeError, "Timecode multiplier cannot be negative" if (arg < 0)
    self.class.new(@total*arg.to_i, @fps)
  end
  
  # Get the next frame
  def succ
    self.class.new(@total + 1, @fps)
  end
  
  # Get the number of times a passed timecode fits into this time span (if performed with Timecode) or 
  # a Timecode that multiplied by arg will give this one
  def /(arg)
    arg.is_a?(Timecode) ?  (@total / arg.total) : self.class.new(@total /arg, @fps)
  end
  
  # Timecodes can be compared to each other
  def <=>(other_tc)
    if framerate_in_delta(fps, other_tc.fps)
      self.total <=> other_tc.total
    else 
      raise WrongFramerate, "Cannot compare timecodes with different framerates"
    end
  end
  
  # FFmpeg expects a fraction of a second as the last element instead of number of frames. Use this
  # method to get the timecode that adheres to that expectation. The return of this method can be fed
  # to ffmpeg directly.
  #  Timecode.parse("00:00:10:24", 25).with_frames_as_fraction #=> "00:00:10.96"
  def with_frames_as_fraction
    vp = value_parts.dup
    vp[-1] = (100.0 / @fps) * vp[-1]
    WITH_FRACTIONS_OF_SECOND % vp
  end
  alias_method :with_fractional_seconds, :with_frames_as_fraction
  
  # Validate that framerates are within a small delta deviation considerable for floats
  def framerate_in_delta(one, two)
    (one.to_f - two.to_f).abs <= ALLOWED_FPS_DELTA
  end
  
  private
  
  # Prepare and format the values for TC output
  def validate!
    frames = @total
    secs = (@total.to_f/@fps).floor
    frames-=(secs*@fps)
    mins = (secs/60).floor
    secs -= (mins*60)
    hrs = (mins/60).floor
    mins-= (hrs*60)

    self.class.validate_atoms!(hrs, mins, secs, frames, @fps)

    [hrs, mins, secs, frames]
  end
  
  def value_parts
    @value ||= validate!
  end
  
end